# 1. Assembly-based variants calling :milky_way:

## Running The Code:
Put the "Aquila/bin" in the ".bashrc" file, and source the ".bashrc" file <br />
Or use the fullpath of "**Aquila_assembly_based_variants_call.py**" and "**Aquila_phasing_all_variants.py**"

```
Aquila/bin/Aquilla_assembly_based_variants_call.py --assembly_dir Assembly_results_S12878 --out_dir Variants_results_S12878 --ref_file Aquila/source/ref.fa 
```
#### *Required parameters
##### --assembly_dir: "Assembly_results_S12878" is the output folder from assembly.
##### --ref_file: "Aquila/source/ref.fa" is the reference fasta file from GRCh38.

#### *Optional parameters
#####  --num_of_threads, default = 1. To increase "--num_of_threads", large memory is needed (see the below Memory/Time Usage table). Maximum is 23 which corresponds to the number of chromosomes.  
#####  --out_dir, default = ./Variants_results

##### --chr_start, --chr_end: if you only want to call variants from some chromosomes or only one chromosome. For example: use "--chr_start 1 --chr_end 5"  will call variants from chromsomes 1,2,3,4,5. Use "--chr_start 2 --chr_end 2" will only call variants from chromosome 2. 

#### *Final Output
##### "Aquila_final_sorted.vcf"

#### Memory/Time Usage
| Memory| --num_of_threads | Time for WGS |
| --- | --- | --- | 
| 100GB | 1 |1-10:24:32 |
| 100GB | 2 | |
| 600GB | 5 | 09:31:08 |

# 2. Phasing all types of variants  :eagle:

```
Aquila/bin/Aquila_phasing_all_variants.py --assembly_vcf Variants_results_S12878/Aquila_final_sorted.vcf --vcf_file S12878_freebayes.vcf --assembly_dir Assembly_results_S12878
```
#### *Required parameters
##### --assembly_vcf: "Variants_results_S12878/Aquila_final_sorted.vcf", it is generated by the above step.
##### --vcf_file: "S12878_freebayes.vcf", it is the origin vcf file generated by FreeBayes. 
##### --assembly_dir: "Assembly_results_S12878" is the output folder from assembly.

#### *Optional parameters
#####  --out_dir, default = ./results_phasing_vcf

##### --chr_start, --chr_end: if you only want to call variants from some chromosomes or only one chromosome. For example: use "--chr_start 1 --chr_end 5"  will call variants from chromsomes 1,2,3,4,5. Use "--chr_start 2 --chr_end 2" will only call variants from chromosome 2. 

#####  --block_len_use, default = 100000, it is the same "--block_len_use" used in assembly steps. 

#### *Final Output
##### "Aquila_all_phased_sorted.vcf"

#### Memory/Time Usage
| Memory | Time for WGS |
| --- | --- | 
| 100GB |05:22:10 |


